sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java

```diff
@@ -301,6 +301,25 @@ public String getId() {
         return withContext(context -> upsertItemInternal(item, requestOptions, context));
     }
 
+    /**
+     * Upserts an item.
+     * <p>
+     * After subscription the operation will be performed. The {@link Mono} upon
+     * successful completion will contain a single resource response with the
+     * upserted item. In case of failure the {@link Mono} will error.
+     *
+     * @param <T> the type parameter.
+     * @param item the item represented as a POJO or Item object to upsert.
+     * @param partitionKey the partition key.
+     * @param options the request options.
+     * @return an {@link Mono} containing the single resource response with the upserted item or an error.
+     */
+    public <T> Mono<CosmosItemResponse<T>> upsertItem(T item, PartitionKey partitionKey, CosmosItemRequestOptions options) {
+        final CosmosItemRequestOptions requestOptions = options == null ? new CosmosItemRequestOptions() : options;
+        ModelBridgeInternal.setPartitionKey(options, partitionKey);
```

Comment:
This is a bug, we need to pass `requestOptions` instead of `options`

Since you are OOF, I will fix it in your branch and push it.